narwhals/_polars/namespace.py

```diff
@@ -219,10 +219,15 @@ def by_dtype(self: Self, dtypes: Iterable[DType]) -> PolarsExpr:
 
         from narwhals._polars.expr import PolarsExpr
 
+        native_dtypes: list[pl.DataType | type[pl.DataType]] = []
+        for dtype in dtypes:
+            native_dtype_instantiated = narwhals_to_native_dtype(dtype, self._version)
+            if isinstance(dtype, type) and issubclass(dtype, DType):
+                native_dtypes.append(native_dtype_instantiated.__class__)
+            else:
+                native_dtypes.append(native_dtype_instantiated)
         return PolarsExpr(
-            pl.selectors.by_dtype(
-                [narwhals_to_native_dtype(dtype, self._version) for dtype in dtypes]
-            ),
```

Comment:
the issue was that we were translating `nw.Datetime` to `pl.Datetime()`. In other parts of the codebase, that's fine. But here, we need to preserve whether the dtype is instantiated or not, as they behave differently

Like this, we keep the translation from `nw.Datetime` to `pl.Datetime`